ext/bcmath/libbcmath/src/div.c

```diff
@@ -31,219 +31,438 @@
+
+		/*
+		 * Determine the temporary quotient.
+		 * The maximum value of numerator_high is divisor_high * B in the previous example, so here numerator_high_part is
+		 * divisor_high_part * BC_VECTOR_BOUNDARY_NUM.
+		 * The number of digits in divisor_high_part is BC_VECTOR_SIZE + 1, so even if divisor_high_part is at its maximum value,
+		 * it will never overflow here.
+		 */
+		numerator_high_part += div_carry * BC_VECTOR_BOUNDARY_NUM * high_part_shift;
+
+		/* numerator_high_part < divisor_high_part => quotient == 0 */
+		if (numerator_high_part < divisor_high_part) {
+			quot_vectors[quot_top_index - i] = 0;
+			div_carry = numerator_vectors[numerator_top_index - i];
+			numerator_vectors[numerator_top_index - i] = 0;
+			continue;
+		}
+
+		BC_VECTOR quot_guess = numerator_high_part / divisor_high_part;
+
+		/* For sub, add the remainder to the high-order digit */
+		numerator_vectors[numerator_top_index - i] += div_carry * BC_VECTOR_BOUNDARY_NUM;
 
-			if (carry != 0) {
-				*rptr = carry;
+		/*
+		 * The temporary quotient can only have errors in the "too large" direction.
+		 * So if the temporary quotient is 0 here, the quotient is 0.
+		 */
+		if (quot_guess == 0) {
+			quot_vectors[quot_top_index - i] = 0;
+			div_carry = numerator_vectors[numerator_top_index - i];
+			numerator_vectors[numerator_top_index - i] = 0;
+			continue;
+		}
+
+		/* sub */
+		BC_VECTOR sub;
+		BC_VECTOR borrow = 0;
+		BC_VECTOR *numerator_calc_bottom = numerator_vectors + numerator_arr_size - divisor_arr_size - i;
+		size_t j;
+		for (j = 0; j < divisor_arr_size - 1; j++) {
+			sub = divisor_vectors[j] * quot_guess + borrow;
+			BC_VECTOR sub_low = sub % BC_VECTOR_BOUNDARY_NUM;
+			borrow = sub / BC_VECTOR_BOUNDARY_NUM;
+
+			if (numerator_calc_bottom[j] >= sub_low) {
+				numerator_calc_bottom[j] -= sub_low;
+			} else {
+				numerator_calc_bottom[j] += BC_VECTOR_BOUNDARY_NUM - sub_low;
+				borrow++;
 			}
 		}
+		/* last digit sub */
+		sub = divisor_vectors[j] * quot_guess + borrow;
+		bool neg_remainder = numerator_calc_bottom[j] < sub;
+		numerator_calc_bottom[j] -= sub;
+
+		/* If the temporary quotient is too large, add back divisor */
+		BC_VECTOR carry = 0;
+		if (neg_remainder) {
+			quot_guess--;
+			for (j = 0; j < divisor_arr_size - 1; j++) {
+				numerator_calc_bottom[j] += divisor_vectors[j] + carry;
+				carry = numerator_calc_bottom[j] / BC_VECTOR_BOUNDARY_NUM;
+				numerator_calc_bottom[j] %= BC_VECTOR_BOUNDARY_NUM;
+			}
+			/* last add */
+			numerator_calc_bottom[j] += divisor_vectors[j] + carry;
+		}
+
+		quot_vectors[quot_top_index - i] = quot_guess;
+		div_carry = numerator_vectors[numerator_top_index - i];
+		numerator_vectors[numerator_top_index - i] = 0;
 	}
 }
 
+static void bc_do_div(const char *n1, size_t n1_readable_len, size_t n1_bottom_extension, const char *n2, size_t n2_len, bc_num *quot, size_t quot_len)
+{
```

Comment:
Would be good to rename n1 and n2 here as well.

Reply:
Renamed
https://github.com/php/php-src/pull/14660/commits/ffb6e8ea3d7142e4082c69b4f241f273f7347048